File: 1337_k_weakest_rows/k_weakest_rows.go
/*
 * @Date: 2021-08-01 15:01:32
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-08-01 15:17:10
 */

package main

import (
	"math/rand"
	"reflect"
	"sort"
	"time"
)

type pair struct{ pow, idx int }

func kWeakestRows(mat [][]int, k int) []int {
	m := len(mat)
	pairs := make([]pair, m)
	for i, row := range mat {
		pow := sort.Search(len(row), func(j int) bool { return row[j] == 0 })
		pairs[i] = pair{pow, i}
	}
	rand.Seed(time.Now().UnixNano())
	randomizedSelected(pairs, 0, m-1, k)
	pairs = pairs[:k]
	sort.Slice(pairs, func(i, j int) bool {
		a, b := pairs[i], pairs[j]
		return a.pow < b.pow || a.pow == b.pow && a.idx < b.idx
	})
	ans := make([]int, k)
	for i, p := range pairs {
		ans[i] = p.idx
	}
	return ans
}

func randomizedSelected(a []pair, l, r, k int) {
	if l >= r {
		return
	}
	pos := randomPartition(a, l, r)
	num := pos - l + 1
	if k == num {
		return
	}
	if k < num {
		randomizedSelected(a, l, pos-1, k)
	} else {
		randomizedSelected(a, pos+1, r, k-num)
	}
}

func randomPartition(a []pair, l, r int) int {
	i := rand.Intn(r-l+1) + l
	a[i], a[r] = a[r], a[i]
	return partition(a, l, r)
}

func partition(a []pair, l, r int) int {
	pivot := a[r]
	i := l - 1
	for j := l; j < r; j++ {
		if a[j].pow < pivot.pow || a[j].pow == pivot.pow && a[j].idx <= pivot.idx {
			i++
			a[i], a[j] = a[j], a[i]
		}
	}
	a[i+1], a[r] = a[r], a[i+1]
	return i + 1
}

func main() {

	assert := func(a, b []int) {
		if !reflect.DeepEqual(a, b) {
			panic("Not Passed")
		}
	}
	{
		mat := [][]int{{1, 1, 0, 0, 0},
			{1, 1, 1, 1, 0},
			{1, 0, 0, 0, 0},
			{1, 1, 0, 0, 0},
			{1, 1, 1, 1, 1}}
		k := 3
		ans := []int{2, 0, 3}
		assert(kWeakestRows(mat, k), ans)
	}
	{
		mat := [][]int{
			{1, 0, 0, 0}, {1, 1, 1, 1}, {1, 0, 0, 0}, {1, 0, 0, 0}}
		k := 2
		ans := []int{0, 2}
		assert(kWeakestRows(mat, k), ans)
	}
}

File: cpp/1337_k_weakest_rows/k_weakest_rows.cpp
/*
 * @Date: 2021-08-01 15:01:23
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-08-01 15:30:04
 */

#include <algorithm>
#include <cassert>
#include <vector>

using namespace std;

template <typename T>
class Helper {
  static int partition(vector<T>& nums, int l, int r) {
    T pivot = nums[r];
    int i = l - 1;
    for (int j = l; j <= r - 1; ++j) {
      if (nums[j] <= pivot) {
        i = i + 1;
        swap(nums[i], nums[j]);
      }
    }
    swap(nums[i + 1], nums[r]);
    return i + 1;
  }

  // 基于随机的划分
  static int randomized_partition(vector<T>& nums, int l, int r) {
    int i = rand() % (r - l + 1) + l;
    swap(nums[r], nums[i]);
    return partition(nums, l, r);
  }

  static void randomized_selected(vector<T>& arr, int l, int r, int k) {
    if (l >= r) {
      return;
    }
    int pos = randomized_partition(arr, l, r);
    int num = pos - l + 1;
    if (k == num) {
      return;
    } else if (k < num) {
      randomized_selected(arr, l, pos - 1, k);
    } else {
      randomized_selected(arr, pos + 1, r, k - num);
    }
  }

 public:
  static vector<T> getLeastNumbers(vector<T>& arr, int k) {
    srand((unsigned)time_t(NULL));
    randomized_selected(arr, 0, (int)arr.size() - 1, k);
    vector<T> vec;
    for (int i = 0; i < k; ++i) {
      vec.push_back(arr[i]);
    }
    return vec;
  }
};

class Solution {
 public:
  vector<int> kWeakestRows(vector<vector<int>>& mat, int k) {
    int m = mat.size(), n = mat[0].size();
    vector<pair<int, int>> power;
    for (int i = 0; i < m; ++i) {
      int l = 0, r = n - 1, pos = -1;
      while (l <= r) {
        int mid = (l + r) / 2;
        if (mat[i][mid] == 0) {
          r = mid - 1;
        } else {
          pos = mid;
          l = mid + 1;
        }
      }
      power.emplace_back(pos + 1, i);
    }

    vector<pair<int, int>> minimum =
        Helper<pair<int, int>>::getLeastNumbers(power, k);
    sort(minimum.begin(), minimum.begin() + k);
    vector<int> ans;
    for (int i = 0; i < k; ++i) {
      ans.push_back(minimum[i].second);
    }
    return ans;
  }
};

int main() {
  {
    vector<vector<int>> mat{{1, 1, 0, 0, 0},
                            {1, 1, 1, 1, 0},
                            {1, 0, 0, 0, 0},
                            {1, 1, 0, 0, 0},
                            {1, 1, 1, 1, 1}};
    int k = 3;
    vector<int> ans{2, 0, 3};
    assert(Solution{}.kWeakestRows(mat, k) == ans);
  }
  {
    vector<vector<int>> mat{
        {1, 0, 0, 0}, {1, 1, 1, 1}, {1, 0, 0, 0}, {1, 0, 0, 0}};
    int k = 2;
    vector<int> ans{0, 2};
    assert(Solution{}.kWeakestRows(mat, k) == ans);
  }
  return 0;
}